File: src/pc.rs
//! Constructs that capture the relevant character-sheet data.
//!
//! This includes ability scores, proficiencies, and other
//! character traits. Rules for how characters interact with
//! the environment (such as darkvision) are also contained
//! here.
use json::JsonValue;
use std::collections::HashMap;
use std::fs::File;
use std::io::Read;
use std::str::FromStr;

use crate::dice::{Dice, Advantage};
use crate::environment::Lighting;

/// Ability score categories
#[derive(Eq, Hash, PartialEq)]
pub enum Ability {
    Strength,
    Dexterity,
    Constitution,
    Intelligence,
    Wisdom,
    Charisma,
}
impl FromStr for Ability {
    type Err = ();
    fn from_str(input: &str) -> Result<Ability, Self::Err> {
        match input {
            "strength" => Ok(Ability::Strength),
            "dexterity" => Ok(Ability::Dexterity),
            "constitution" => Ok(Ability::Constitution),
            "intelligence" => Ok(Ability::Intelligence),
            "wisdom" => Ok(Ability::Wisdom),
            "charisma" => Ok(Ability::Charisma),
            _ => Err(()),
        }
    }
}

/// Things which one can be proficient in
#[derive(Eq, Hash, PartialEq)]
pub enum Proficiency {
    Insight,
    Investigation,
    Perception,
}
impl FromStr for Proficiency {
    type Err = ();
    fn from_str(input: &str) -> Result<Proficiency, Self::Err> {
        match input {
            "insight"  => Ok(Proficiency::Insight),
            "investigation"  => Ok(Proficiency::Investigation),
            "perception"  => Ok(Proficiency::Perception),
            _ => Err(()),
        }
    }
}

/// A player character.
///
/// This struct is roughly equivalent to a dndbeyond character
/// sheet, in that it encapsulates the rules for rolling for
/// various checks alongside the values that numerically affect
/// the outcome.
pub struct PC<'a> {
    /// The source of randomness that a character uses to make rolls.
    dice: &'a Dice,
    /// A mapping from ability to ability score.
    abilities: HashMap<Ability, u8>,
    /// A mapping from proficiency to proficiency level.
    ///
    /// When making a roll, the PC multiplies their proficiency bonus
    /// by their proficiency level (0, 1, or 2) to determine the
    /// proficiency modifier.
    proficiencies: HashMap<Proficiency, u8>,
    /// The PC's proficiency bonus.
    proficiency_bonus: u8,
    /// Whether the PC has darkvision.
    darkvision: bool,
}

impl PC<'_> {
    /// Create a character from a configuration file.
    ///
    /// * `dice` - The dice that the PC will use to generate randomness.
    /// * `config` - A JSON configuration file which lays out the character's attributes.
    pub fn new<'a, 'b>(dice: &'a Dice, config: &'b mut File) -> std::io::Result<PC<'a>> {
        let mut abilities: HashMap<Ability, u8> = HashMap::new();
        let mut proficiencies: HashMap<Proficiency, u8> = HashMap::new();

        // Parse the file to a JSON value
        let mut config_string: String = String::new();
        config.read_to_string(&mut config_string)?;
        let config_data: JsonValue = match json::parse(&config_string) {
            Ok(data) => data,
            Err(_error) => panic!(),
        };

        // Insert each ability score into the PC's abilities.
        // Panic if they don't coerce to the enum or if their values
        // don't fit in a u8.
        for (ability, score) in config_data["abilities"].entries() {
            let ability_val: Ability = match Ability::from_str(ability) {
                Ok(data) => data,
                Err(_error) => panic!(),
            };
            let score_val: u8 = match score.as_u8() {
                Some(data) => data,
                None => panic!(),
            };
            abilities.insert(ability_val, score_val);
        }

        // Insert each proficiency into the PC's proficiencies.
        // Panic if they don't coerce to the enum or if their values
        // don't fit in a u8.
        for (proficiency, level) in config_data["proficiencies"].entries() {
            let proficiency_val: Proficiency = match Proficiency::from_str(proficiency) {
                Ok(data) => data,
                Err(_error) => panic!(),
            };
            let level_val: u8 = match level.as_u8() {
                Some(data) => data,
                None => panic!(),
            };
            proficiencies.insert(proficiency_val, level_val);
        }

        // Set the proficiency bonus.
        let proficiency_bonus: u8 = match config_data["proficiency_bonus"].as_u8() {
            Some(data) => data,
            None => panic!(),
        };

        // Set the darkvision status.
        let darkvision: bool = match config_data["darkvision"].as_bool() {
            Some(data) => data,
            None => panic!(),
        };

        Ok(PC {dice, abilities, proficiencies, proficiency_bonus, darkvision})
    }

    /// Given a profiency category, return the PC's proficiency modifier.
    ///
    /// The proficiency modifier is the PC's proficiency bonus multiplied
    /// by their level of proficiency (0, 1, or 2 for expertise).
    /// * `proficiency` - The type of proficiency whose modifier to retrieve.
    pub fn proficiency_modifier(&self, proficiency: Proficiency) -> u8 {
        self.proficiency_bonus * self.proficiencies[&proficiency]
    }

    /// Given an ability score, return the PC's ability modifier.
    /// * `ability` - The ability whose modifier to retrieve.
    pub fn ability_modifier(&self, ability: Ability) -> i8 {
        let ability_score: u8 = self.abilities[&ability];
        (ability_score as i8 - 10) / 2
    }

    /// Roll a check.
    ///
    /// Roll a d20, adding the appropriate ability and proficiency modifiers,
    /// and with the appropriate level of advantage.
    /// * `ability` - The ability to apply to the check.
    /// * `proficiency` - The proficiency to apply to the check.
    /// * `advantage` - The advantage level of the check.
    pub fn check(&mut self, ability: Ability, proficiency: Proficiency, advantage: Advantage) -> i8 {
        let proficiency_bonus: u8 = self.proficiency_modifier(proficiency);
        let ability_score: i8 = self.ability_modifier(ability);
        let total_modifier: i8 = proficiency_bonus as i8 + ability_score;

        self.dice.d(20, total_modifier, advantage)
    }

    /// Roll a Wisdom (Perception) check.
    ///
    /// Apply all available modifiers, including potential disadvantage from
    /// lighting conditions.
    /// * `advantage` - Any additional advantage beyond the usual perception parameters.
    /// * `lighting` - The level of environmental lighting.
    pub fn perception_check(&mut self, advantage: Advantage, lighting: Lighting) {
        let lighting_advantage: Advantage = match lighting {
            Lighting::Dark => if self.darkvision { Advantage::Disadvantage } else { Advantage::Fail },
            Lighting::Dim => if self.darkvision { Advantage::None } else { Advantage::Disadvantage },
            Lighting::Light => Advantage::None,
        };
        self.check(Ability::Wisdom, Proficiency::Perception, advantage + lighting_advantage);
    }
}

File: pygrue/pc.py
from __future__ import annotations
from collections import defaultdict
import enum
import json
import math
import pathlib

from .dice import Advantage, Dice
from .environment import Lighting


class Ability(enum.Enum):
    """
    Enumerated ability scores.
    """
    STRENGTH = enum.auto()
    DEXTERITY = enum.auto()
    CONSTITUTION = enum.auto()
    INTELLIGENCE = enum.auto()
    WISDOM = enum.auto()
    CHARISMA = enum.auto()

    @staticmethod
    def from_str(source: str) -> Ability:
        """
        Convert a string to an `Ability` enum. Returns the appropriate enum
        value for the source string, where the string must be the full name
        of the ability, case insensitive.

        :params source: The string to convert
        """
        for value in Ability:
            if value.name == source.upper():
                return value
        raise ValueError(f"Cannot interpret {source} as Ability")


class Proficiency(enum.Enum):
    """
    Enumerated possible proficienies.
    """
    INSIGHT = enum.auto()
    INVESTIGATION = enum.auto()
    PERCEPTION = enum.auto()
    STEALTH = enum.auto()

    @staticmethod
    def from_str(source: str) -> Proficiency:
        """
        Convert a string to a `Proficiency` enum. Returns the appropriate enum
        value for the source string, where the string must be the full name
        of the proficiency, case insensitive.

        :params source: The string to convert
        """
        for value in Proficiency:
            if value.name == source.upper():
                return value
        raise ValueError(f"Cannot interpret {source} as Proficiency")

    def sight_affected(self) -> bool:
        """
        Returns true if rolls using this proficiency can be affected by line of
        sight and lighting conditions.
        """
        return self in [
            Proficiency.INSIGHT,
            Proficiency.INVESTIGATION,
            Proficiency.PERCEPTION,
        ]


class ProficiencyLevel(enum.IntEnum):
    """
    Proficiency levels, as an integer. Multiplying this value by a character's
    proficiency bonus will result in the correct modifier for the given
    proficiency.
    """
    NONE = 0
    PROFICIENCY = 1
    EXPERTISE = 2


class PC:
    """
    A character with all relevant attributes for calculating behind-the-screen
    ability checks.
    """
    __slots__ = {
        "abilities": "A `dict` between `Ability`s and modifiers",
        "name": "The character name",
        "proficiencies": "A `defaultdict` between `Proficiency`s and `ProficiencyLevel`s (default 0)",
        "proficiency_advantages": "A `defaultdict` between `Proficiency`s and `Advantage`s (default `NONE`)",
        "proficiency_bonus": "The integer proficiency bonus the character uses",
        "darkvision": "`True` if the character has darkvision, false otherwise",
    }

    def __init__(self, character_sheet: pathlib.Path):
        """
        :params character_sheet: A path to a JSON-encoded character sheet
        """
        with open(character_sheet, 'r') as sheet:
            data = json.load(sheet)
            self.name = data["name"]

            self.abilities = {
                Ability.from_str(name): PC._modifier_from_score(value)
                for name, value in data["abilities"].items()
            }

            self.proficiencies = defaultdict(lambda: 0)
            for name, value in data["proficiencies"].items():
                self.proficiencies[Proficiency.from_str(name)] = ProficiencyLevel(value)

            self.proficiency_advantages = defaultdict(lambda: Advantage.NONE)
            for name, value in data["proficiency_advantages"].items():
                match value:
                    case 0:
                        self.proficiency_advantages[Proficiency.from_str(name)] = Advantage.NONE
                    case 1:
                        self.proficiency_advantages[Proficiency.from_str(name)] = Advantage.ADVANTAGE
                    case -1:
                        self.proficiency_advantages[Proficiency.from_str(name)] = Advantage.DISADVANTAGE
                    case _:
                        raise ValueError(f"Proficiency advantages must be from -1, 0, or 1, not {value}")

            self.proficiency_bonus = data["proficiency_bonus"]
            self.darkvision = data["darkvision"]

        assert all(a in self.abilities for a in Ability), f"Not all abilities populated for {self.name}"

    @staticmethod
    def _modifier_from_score(score: int) -> int:
        """
        Convert an ability score to the corresponding modifier
        (i.e. 8 -> -1, 11 -> 1, etc.)

        :params score: The ability score (usually from 1-20)
        """
        diff = score - 10
        return math.floor(diff / 2)

    def check(self,
              ability: Ability,
              proficiency: Proficiency | None = None,
              advantage: Advantage = Advantage.NONE) -> int:
        """
        Perform an ability check, with optional proficiency and advantage
        modifiers.

        :params ability: The ability to use for the check
        :params proficiency: A proficiency to apply to the check (by default, no proficiency applies)
        :params advantage: The advantage state to apply to the check
        """
        roll = Dice.d(20)
        roll.modify(self.abilities[ability] + self.proficiencies[proficiency] * self.proficiency_bonus)
        total_advantage = self.proficiency_advantages[proficiency] + advantage
        result = roll.apply(total_advantage)
        return result

    def with_lighting(self, lighting: Lighting) -> Advantage:
        """
        Return the advantage state of the character's sight-based checks under
        the given lighting condition, based on the character's darkvision.

        :params lighting: The lighting condition under which to sense
        """
        if lighting is Lighting.BRIGHT:
            return Advantage.NONE
        elif lighting is Lighting.DIM:
            return Advantage.NONE if self.darkvision else Advantage.DISADVANTAGE
        elif lighting is Lighting.DARK:
            return Advantage.DISADVANTAGE if self.darkvision else Advantage.FAIL
        else:
            raise NotImplementedError(f"Lighting type {lighting.name} not supported.")

    def sight_based_check(self,
                          ability: Ability,
                          lighting: Lighting,
                          proficiency: Proficiency | None = None,
                          advantage: Advantage = Advantage.NONE) -> int:
        """
        Perform an ability check, with optional proficiency and advantage
        modifiers, applying the intrinsic advantage state the results from
        the lighting conditions and the character's darkvision.

        :params ability: The ability to use for the check
        :params lighting: The lighting conditions under which the check is made
        :params proficiency: A proficiency to apply to the check, before lighting conditions (by default, none)
        :params advantage: The advantage state to apply to the check
        """
        total_advantage = advantage + self.with_lighting(lighting)
        return self.check(ability, proficiency, total_advantage)
